src/vivarium/framework/engine.py

```diff
@@ -35,107 +33,139 @@
-from .lifecycle import LifeCycleInterface
+from .lifecycle import LifeCycleInterface, LifeCycleError
 
 _log = logging.getLogger(__name__)
 
 
 class SimulationContext:
-    """context"""
-    def __init__(self, configuration, components, plugin_manager=None):
-        plugin_manager = plugin_manager if plugin_manager else PluginManager()
 
-        self.configuration = configuration
+    def __init__(self, model_specification: Union[str, Path, ConfigTree] = None,
+                 components: Union[List, Dict[str, str], ConfigTree] = None,
+                 configuration: Union[Dict, ConfigTree] = None,
+                 plugin_configuration: Union[Dict[str, str], ConfigTree] = None):
+        # Bootstrap phase: Parse arguments, make private managers
+        component_configuration = components if isinstance(components, (dict, ConfigTree)) else None
+        self._additional_components = components if isinstance(components, List) else []
+        model_specification = build_model_specification(model_specification, component_configuration,
+                                                        configuration, plugin_configuration)
 
-        self.lifecycle = plugin_manager.get_plugin('lifecycle')
-        self.lifecycle.add_phase('initialization', ['initialization'])
-        self.lifecycle.add_phase('setup', ['setup', 'post_setup', 'population_creation'])
-        self.lifecycle.add_phase(
-            'main_loop', ['time_step__prepare', 'time_step', 'time_step__cleanup', 'collect_metrics'], loop=True
-        )
-        self.lifecycle.add_phase('simulation_end', ['simulation_end'])
-        self.lifecycle.set_state('initialization')
+        self._plugin_configuration = model_specification.plugins
+        self._component_configuration = model_specification.components
+        self.configuration = model_specification.configuration
 
-        self.component_manager = plugin_manager.get_plugin('component_manager')
-        self.component_manager.configuration = self.configuration
+        self._plugin_manager = PluginManager(model_specification.plugins)
 
-        # TODO: remove when lifecycle restrictions are implemented.
-        self._setup = False
+        # TODO: Setup logger here.
+
+
+        self._builder = Builder(self.configuration, self._plugin_manager)
+
+        # This formally starts the initialization phase (this call makes the
+        # life-cycle manager).
+        self._lifecycle = self._plugin_manager.get_plugin('lifecycle')
+        self._lifecycle.add_phase('setup', ['setup', 'post_setup', 'population_creation'])
+        self._lifecycle.add_phase(
+            'main_loop', ['time_step__prepare', 'time_step', 'time_step__cleanup', 'collect_metrics'], loop=True
+        )
+        self._lifecycle.add_phase('simulation_end', ['simulation_end', 'report'])
 
-        self.builder = Builder(configuration, plugin_manager)
+        self._component_manager = self._plugin_manager.get_plugin('component_manager')
```

Comment:
made all the manager private.  Otherwise this code is essentially the same as before.